packages/frontend/core/src/components/workspace-upgrade/upgrade-hooks.ts

```diff
@@ -1,42 +1,126 @@
-import { forceUpgradePages } from '@toeverything/infra/blocksuite';
-import { useCallback, useState } from 'react';
+import { WorkspaceFlavour } from '@affine/env/workspace';
+import type { RootWorkspaceMetadata } from '@affine/workspace/atom';
+import { getOrCreateWorkspace } from '@affine/workspace/manager';
+import type { Workspace } from '@blocksuite/store';
+import { useAsyncCallback } from '@toeverything/hooks/affine-async-hooks';
+import { getBlockSuiteWorkspaceAtom } from '@toeverything/infra/__internal__/workspace';
+import { getCurrentStore } from '@toeverything/infra/atom';
+import type { MigrationPoint } from '@toeverything/infra/blocksuite';
+import {
+  migrateLocalBlobStorage,
+  migrateWorkspace,
+} from '@toeverything/infra/blocksuite';
+import { nanoid } from 'nanoid';
+import { useState } from 'react';
+import { applyUpdate, Doc as YDoc, encodeStateAsUpdate } from 'yjs';
 
+import { WorkspaceAdapters } from '../../adapters/workspace';
 import { useCurrentSyncEngine } from '../../hooks/current/use-current-sync-engine';
 import { useCurrentWorkspace } from '../../hooks/current/use-current-workspace';
 
 export type UpgradeState = 'pending' | 'upgrading' | 'done' | 'error';
 
-export function useUpgradeWorkspace() {
+function applyDoc(target: YDoc, result: YDoc) {
+  applyUpdate(target, encodeStateAsUpdate(result));
+  for (const targetSubDoc of target.subdocs.values()) {
+    const resultSubDocs = Array.from(result.subdocs.values());
+    const resultSubDoc = resultSubDocs.find(
+      item => item.guid === targetSubDoc.guid
+    );
+    if (resultSubDoc) {
+      applyDoc(targetSubDoc, resultSubDoc);
+    }
+  }
+}
+
+export function useUpgradeWorkspace(migration: MigrationPoint) {
   const [state, setState] = useState<UpgradeState>('pending');
   const [error, setError] = useState<Error | null>(null);
+  const [newWorkspaceId, setNewWorkspaceId] = useState<string | null>(null);
 
   const [workspace] = useCurrentWorkspace();
   const syncEngine = useCurrentSyncEngine();
+  const rootStore = getCurrentStore();
 
-  const upgradeWorkspace = useCallback(() => {
+  const upgradeWorkspace = useAsyncCallback(async () => {
     setState('upgrading');
     setError(null);
+    try {
+      if (environment.isDesktop) {
```

Comment:
I'm not so sure,
It seems that Decktop starting stage upgrade only handles local workspace? cloud workspace may still need this function.

Reply:
Agreed. Assume on the desktop the user enabled cloud and he opens a never opened outdated cloud workspace, the data will be pulled from remote and will still go into this function.

Reply:
Good point! This condition can be removed!